Automattic/jetpack
# CRM: Making sure invoice total count factors in credit notes and refunds correctly in invoice editor, preview and pdf view

Issue: https://github.com/Automattic/zero-bs-crm/issues/2868

## Proposed changes:

* This PR primarily changes the logic in `ZeroBSCRM.InvoiceBuilder.php` and `ZeroBSCRM.DAL3.Obj.Invoices.php` in terms of how the invoice total is calculated and displayed when refunds and credit notes are factored in.
* This fixes the issue where refunds and credit notes were being incorrectly added to totals when previewing or viewing the pdf of an invoice, instead of being subtracted. However this 'fix' may itself need further scrutiny - it assumes that if a credit or refund is added (either minus or positive), it is meant to be subtracted from the invoice total.
* This PR still separates out Sales from Refunds / Credits due to the current behaviour of Sales being a little confusing (see below).
* Also, I did initially approach this issue by making sure that the preview and pdf pages matched the invoice edit page. What I've since realized is that fixing that would be a breaking change for people - if they had been inputting credits and refunds they would most likely have been entering them as negative amounts because the Client Portal and PDF are what matter most, as that is shared with customers. So the proposed fix is a catch-all for refunds and credits, and also allows further conversation about best practices.


### Existing quirks / related behaviour for comparison

When testing out totals I discovered some confusing behaviour (to me). This is likely expected, though without being able to grasp the logic behind it I'm noting some of the behaviour here, partly because anyone testing this PR may look at some of this behaviour and possibly be confused about expected vs new behaviour:

* When combined with WooCommerce and using WooSync and then setting WooSync settings to create invoices from orders, those invoices will have an associated transaction created that effectively pays off the invoice (assuming you've set transaction settings to include Pending statuses in the total value from the CRM transactions settings page). The transaction will be marked as a sale and have a positive value (matching the invoice total). 
* If you then apply a partial refund to that order within WooCommerce, that refund will generate a new transaction with a negative value sale. However that transaction is not automatically linked to an invoice. Linking it to the original invoice will only temporary work - each WooSync will unlink it. Linking it to the original invoice will however increase the invoice total.
* So, in effect, it appears that both positive and negative sales values look like they should pay off an invoice value given the current behaviour.
* I've shared more about this in Slack - p1700570444797999-slack-CTXBP902X - and now also in this issue - https://github.com/Automattic/zero-bs-crm/issues/3386. 
* For now however, I've approached this just by fixing what the seemingly current expected behaviour would be.

### Other information:


## Jetpack product discussion

Issue: https://github.com/Automattic/zero-bs-crm/issues/2868

## Does this pull request change what data or activity we track or use?

No.

## Testing instructions:

* Using the [Jetpack Beta plugin](https://github.com/Automattic/jetpack-beta) on a test site with this branch applied or testing locally with the branch checked out, create a new invoice assigned to a contact, with a product and value greater than 0.
* Create a new transaction, set the transaction value to be less than the invoice value, and assign it to the contact and invoice just created. Test with all three types: Sale, Refund and Credit Note (when manually inputting, you can add either negative or positive for refund and credit - they should behave the same. With sales, the behaviour should match what it did before - positive value decreases the total amount, negative value increases it). See more in the existing quirks section for what may need to change in the future).
* On the invoice edit page, check the totals are correct. Then click Preview as well as downloading the pdf of the invoice - the totals should be correct (the total amount reduced for all three).
